### DISCUSSION TITLE
Working with virtual environments (Python LSP)

### DISCUSSION THREAD

**Discussion by pdelboca** [orbitalquark/textadept#527] [Q&A]
Hello all!

First of all, thanks for doing Textadept! I'm exciting to give it a try!

I work mostly with python and I'm trying to configure the LSP server to take into consideration my virtual environments instead of the global Python installation.

I'm currently using [pylsp](https://github.com/python-lsp/python-lsp-server) and the default configuration works as expected:
```
-- LSP
local lsp = require('lsp')
lsp.log_rpc = true
lsp.server_commands.python = 'pylsp'
```

However, I would like the LSP to work with the python and the packages I have in my virtual enviroments. On `nvim`  I solve this by just activating it before executing `nvim .` to open the project.

Any suggestions on how can I achieve a similar result with Textadept? I always use a `.venv` folder in my project (next to my `.git` folder), so I'm wondering if there is any way to code something in my `init.lua` file.

Thanks!

**Comment by orbitalquark** [comment]
You can try activating your virtual environment and then running `textadept` from the command line (possibly with the `-f` or `--force` flag if there's already an open instance), sort of like how you did it with nvim. Textadept will inherit the proper environment variables, and the spawned LSP server should also inherit them.

Otherwise, you may have to pass flags to the 'pylsp' server command.

**Comment by pdelboca** [comment]
Hello @orbitalquark . Thanks for your response.

When I activate a virtual environment before executing `textadept -f` it fails to execute basic LSP operations like `Go To Definition`.

It looks like it might be a problem with `pyslp` so I will ask in their repository.

## More context for the curious

**Note:** I'm using `python -m venv .venv` to create my virtual environments.

Checking at the logs, the initialization of the LSP Server sends/receive the exact same json. (Checked it with https://www.jsondiff.com/)

Executing `$ textadept -f`:
```
Starting language server: pylsp
RPC send: {"params":{"capabilities":{"textDocument":{"hover":{"contentFormat":["plaintext"]},"publishDiagnostics":[],"documentSymbol":{"symbolKind":{"valueSet":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26]}},"signatureHelp":{"signatureInformation":{"activeParameterSupport":true,"documentationFormat":["plaintext"],"parameterInformation":{"labelOffsetSupport":true}}},"completion":{"completionItem":{"snippetSupport":true,"preselectSupport":true,"documentationFormat":["plaintext"]},"completionItemKind":{"valueSet":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25]}},"synchronization":{"didSave":true}}},"clientInfo":{"name":"textadept","version":"Textadept 12.4"},"processId":null,"rootUri":"file:///home/pdelboca/Repos/anagrafeboca"},"jsonrpc":"2.0","id":1,"method":"initialize"}
RPC recv: {"jsonrpc":"2.0","id":1,"result":{"capabilities":{"codeActionProvider":true,"codeLensProvider":{"resolveProvider":false},"completionProvider":{"resolveProvider":true,"triggerCharacters":["."]},"documentFormattingProvider":true,"documentHighlightProvider":true,"documentRangeFormattingProvider":true,"documentSymbolProvider":true,"definitionProvider":true,"executeCommandProvider":{"commands":[]},"hoverProvider":true,"referencesProvider":true,"renameProvider":true,"foldingRangeProvider":true,"signatureHelpProvider":{"triggerCharacters":["(",",","="]},"textDocumentSync":{"change":2,"save":{"includeText":true},"openClose":true},"notebookDocumentSync":{"notebookSelector":[{"cells":[{"language":"python"}]}]},"workspace":{"workspaceFolders":{"supported":true,"changeNotifications":true}},"experimental":{}},"serverInfo":{"name":"pylsp","version":"1.11.0"}}}
Connected to pylsp 1.11.0
```

Executing `(.venv)$ textadept -f`
```
Starting language server: pylsp
RPC send: {"method":"initialize","jsonrpc":"2.0","params":{"rootUri":"file:///home/pdelboca/Repos/anagrafeboca","capabilities":{"textDocument":{"completion":{"completionItem":{"documentationFormat":["plaintext"],"snippetSupport":true,"preselectSupport":true},"completionItemKind":{"valueSet":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25]}},"documentSymbol":{"symbolKind":{"valueSet":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26]}},"publishDiagnostics":[],"hover":{"contentFormat":["plaintext"]},"signatureHelp":{"signatureInformation":{"documentationFormat":["plaintext"],"activeParameterSupport":true,"parameterInformation":{"labelOffsetSupport":true}}},"synchronization":{"didSave":true}}},"clientInfo":{"version":"Textadept 12.4","name":"textadept"},"processId":null},"id":1}
RPC recv: {"jsonrpc":"2.0","id":1,"result":{"capabilities":{"codeActionProvider":true,"codeLensProvider":{"resolveProvider":false},"completionProvider":{"resolveProvider":true,"triggerCharacters":["."]},"documentFormattingProvider":true,"documentHighlightProvider":true,"documentRangeFormattingProvider":true,"documentSymbolProvider":true,"definitionProvider":true,"executeCommandProvider":{"commands":[]},"hoverProvider":true,"referencesProvider":true,"renameProvider":true,"foldingRangeProvider":true,"signatureHelpProvider":{"triggerCharacters":["(",",","="]},"textDocumentSync":{"change":2,"save":{"includeText":true},"openClose":true},"notebookDocumentSync":{"notebookSelector":[{"cells":[{"language":"python"}]}]},"workspace":{"workspaceFolders":{"supported":true,"changeNotifications":true}},"experimental":{}},"serverInfo":{"name":"pylsp","version":"1.11.0"}}}
Connected to pylsp 1.11.0
```

However, when sending a Go To Definition to the LSP, it returns nothing when the `venv` is activated. While it returns correctly if I do not have it activated.

Executing `$ textadept -f`

```
RPC send: {"params":{"position":{"character":24,"line":12},"textDocument":{"uri":"file:///home/pdelboca/Repos/anagrafeboca/anagrafeboca/settings.py"}},"jsonrpc":"2.0","id":2,"method":"textDocument/definition"}
RPC recv: {"jsonrpc":"2.0","id":2,"result":[{"uri":"file:\/\/\/usr\/local\/lib\/python3.10\/pathlib.py","range":{"start":{"line":942,"character":6},"end":{"line":942,"character":10}}}]}
RPC recv: {"jsonrpc":"2.0","method":"textDocument\/publishDiagnostics","params":{"uri":"file:\/\/\/home\/pdelboca\/Repos\/anagrafeboca\/anagrafeboca\/settings.py","diagnostics":[]}}
```

Executing `(.venv) $ textadept -f`

```
RPC send: {"method":"textDocument/definition","jsonrpc":"2.0","params":{"position":{"character":24,"line":12},"textDocument":{"uri":"file:///home/pdelboca/Repos/anagrafeboca/anagrafeboca/settings.py"}},"id":2}
RPC recv: {"jsonrpc":"2.0","id":2,"result":[]}
RPC recv: {"jsonrpc":"2.0","method":"textDocument\/publishDiagnostics","params":{"uri":"file:\/\/\/home\/pdelboca\/Repos\/anagrafeboca\/anagrafeboca\/settings.py","diagnostics":[]}}
```

**Comment by pdelboca** [comment]
Hello!

Just to circle back here. I think what was causing the error was an issue in my path. I think and old instalation of `pipx` was messing with it. I cleaned it and now activating the virtual environment before executing `$ textadept` seems to work just fine.

Also, I read about [how to configure environments](https://github.com/python-lsp/python-lsp-server/issues/496#issuecomment-1839839544) in `pylsp`. Seems that a configuration like the following one should be enough (although it didn't work for me):

```lua
-- LSP
local lsp = require('lsp')
lsp.log_rpc = true
lsp.server_commands.python = function()
    config = {
      pylsp = {
        plugins = {
          jedi = {
            environment = './.venv/'
          },
        }
      }
    }
    return 'pylsp', config
  end

```

So, long story short: activating the virtual environment before running textadept should do the trick.

Thanks!
